### PR TITLE
Added YAML Parser

### PR DESCRIPTION
**Pull request by maxwaldman8**

### CODE REVIEW

File: AstraEngine.Serialization/AstraEngine.Serialization.csproj [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
```
Comment (jcollard) [inline AstraEngine.Serialization/AstraEngine.Serialization.csproj:4]: This should not produce an executable. You should remove this line.

File: AstraEngine.Serialization/AstraEngine.Serialization.csproj [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
```
Comment (jcollard) [inline AstraEngine.Serialization/AstraEngine.Serialization.csproj:3]: @maxwaldman8 
When you intend other people to use your code as a library, you should include this line in the `<PropertyGroup>` section: `<GenerateDocumentationFile>true</GenerateDocumentationFile>`.

You can see this here:

https://github.com/maxwaldman8/AstraEngine/blob/3-Feature-Request-YAML-Parser/AstraEngine.InputSystem/AstraEngine.InputSystem.csproj

```xml
﻿<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
  </PropertyGroup>

</Project>
```

File: AstraEngine.Serialization/Parser.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿using System.Reflection;
+using YamlDotNet.Core.Events;
+using YamlDotNet.Serialization;
+
+namespace AstraEngine.Serialization;
+
+public class Parser {
```
Comment (jcollard) [inline AstraEngine.Serialization/Parser.cs:7]: @maxwaldman8  I'd recommend making this a static class.

File: AstraEngine.Serialization/Parser.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    public bool Resolve(NodeEvent? nodeEvent, ref Type currentType) {
+        try {
+            if (nodeEvent!.Tag.ToString()[0] == '!') {
+                currentType = Type.GetType(
+                    Type.GetType(nodeEvent!.Tag.ToString()[1..]) == null
+                    ? $"{nodeEvent!.Tag.ToString()[1..]}, {Assembly.GetEntryAssembly()}"
+                    : nodeEvent!.Tag.ToString()[1..]
+                )!;
+            }
+        } catch {
```
Comment (jcollard) [inline AstraEngine.Serialization/Parser.cs:21]: It is typically bad practice to not specify the exception type. What exception are you expecting to catch here?

**Review by jcollard — CHANGES_REQUESTED** [review]
You should provide unit tests showing how to use this as well as a use case in the `AstraEngine.Examples`.

**Review by jcollard — CHANGES_REQUESTED** [review]
A Herculean effort here! 

Some small things to fix / consider before this is merged.

File: AstraEngine.Core/Component.cs [hunk 5]
```diff
@@ -13,9 +13,19 @@ public abstract class Component
...
     [AllowNull]
     public Entity Entity { get; internal set; }
+    /// <summary>Whether or not this <see cref="Component"/> is active</summary>
+    public bool IsActive { get; set; } = true;
 
     /// <summary>
-    /// Executed when this component enters the game for the first time
+    /// Executed when the game first starts
+    /// </summary>
+    public void Start()
```
[the remark this answers is not in the document]
Comment (maxwaldman8) [reply AstraEngine.Core/Component.cs:22]: Shouldn't this change (and the other ones below this) be for my other pull request? I am not sure why they appeared as changed files on this one.

[the remark this answers is not in the document]
Comment (jcollard) [reply AstraEngine.Core/Component.cs:22]: You may need to merge `GameEngineDesign-S24:main` into your branch.

File: AstraEngine.Core/Entity.cs [hunk 6]
```diff
@@ -269,6 +286,7 @@ public bool AddChild(Entity child)
     {
         if (_children.Add(child))
         {
+            child.Initialize();
```
Comment (jcollard) [inline AstraEngine.Core/Entity.cs:289]: What do you do if the parent is not active when the child is added? What if the child being added is not active?

**Review by jcollard — COMMENTED** [review]
I'm not sure remove Initialize from Tick is the best thing to do. Let's talk about it first.

**Review by jcollard — APPROVED** [review]
Looking good! Fantastic effort @maxwaldman8

**Pull request merged**